File: scanner/favicon.py
# -*- coding: utf-8 -*-
"""favicon 指纹：fofa 风格 icon_hash（mmh3(base64(icon)))。

TscanPlus 指纹中的 icon_hash 条件即此算法；mmh3 用纯 Python 实现
（32 位 x86 MurmurHash3），避免额外依赖。
"""
import base64


def mmh3_32(data, seed=0):
    """MurmurHash3 x86 32 位（返回有符号整数，与 fofa 一致）"""
    data = bytearray(data)
    length = len(data)
    nblocks = length // 4
    c1 = 0xCC9E2D51
    c2 = 0x1B873593
    h = seed & 0xFFFFFFFF

    def rotl(x, r):
        return ((x << r) | (x >> (32 - r))) & 0xFFFFFFFF

    for i in range(nblocks):
        k = (data[i * 4] | (data[i * 4 + 1] << 8) |
             (data[i * 4 + 2] << 16) | (data[i * 4 + 3] << 24))
        k = (k * c1) & 0xFFFFFFFF
        k = rotl(k, 15)
        k = (k * c2) & 0xFFFFFFFF
        h ^= k
        h = rotl(h, 13)
        h = (h * 5 + 0xE6546B64) & 0xFFFFFFFF

    tail = data[nblocks * 4:]
    k = 0
    if len(tail) >= 3:
        k ^= tail[2] << 16
    if len(tail) >= 2:
        k ^= tail[1] << 8
    if len(tail) >= 1:
        k ^= tail[0]
        k = (k * c1) & 0xFFFFFFFF
        k = rotl(k, 15)
        k = (k * c2) & 0xFFFFFFFF
        h ^= k

    h ^= length
    h ^= h >> 16
    h = (h * 0x85EBCA6B) & 0xFFFFFFFF
    h ^= h >> 13
    h = (h * 0xC2B2AE35) & 0xFFFFFFFF
    h ^= h >> 16

    # 转有符号 32 位
    if h >= 0x80000000:
        h -= 0x100000000
    return h


def favicon_hash(icon_bytes):
    """fofa icon_hash：base64 编码后取 mmh3"""
    b64 = base64.encodebytes(icon_bytes)
    return mmh3_32(b64)

File: internal/sitelens/favicon.go
// favicon mmh3 指纹：FOFA 生态的 icon_hash 算法——内容 base64 编码
// （codecs.encode 语义，含结尾换行）后取 murmur3 x86_32（seed 0），
// 按有符号 int32 语义返回。算法已与官方 mmh3 C 实现逐位锚定
// （校准脚本 build/mm3_ref.py；测试内含锚定向量）。
package sitelens

import (
	"encoding/base64"
	"math/bits"
	"strings"

	"cnb.cool/feng-qiao/sitelens/internal/httpx"
)

// murmur3x8632 MurmurHash3 x86 32 位（seed 固定 0）。
func murmur3x8632(data []byte) uint32 {
	const c1, c2 = 0xcc9e2d51, 0x1b873593
	h := uint32(0)
	n := len(data) / 4
	for i := 0; i < n; i++ {
		k := uint32(data[i*4]) | uint32(data[i*4+1])<<8 |
			uint32(data[i*4+2])<<16 | uint32(data[i*4+3])<<24
		k *= c1
		k = bits.RotateLeft32(k, 15)
		k *= c2
		h ^= k
		h = bits.RotateLeft32(h, 13)
		h = h*5 + 0xe6546b64
	}
	var k uint32
	tail := data[n*4:]
	if len(tail) >= 3 {
		k ^= uint32(tail[2]) << 16
	}
	if len(tail) >= 2 {
		k ^= uint32(tail[1]) << 8
	}
	if len(tail) >= 1 {
		k ^= uint32(tail[0])
		k *= c1
		k = bits.RotateLeft32(k, 15)
		k *= c2
		h ^= k
	}
	h ^= uint32(len(data))
	h ^= h >> 16
	h *= 0x85ebca6b
	h ^= h >> 13
	h *= 0xc2b2ae35
	h ^= h >> 16
	return h
}

// fofaBase64 复刻 Python codecs.encode(content, "base64")（即
// base64.encodebytes）：76 字符换行、每行以 \n 结尾、空输入输出空。
func fofaBase64(content []byte) []byte {
	if len(content) == 0 {
		return nil
	}
	enc := base64.StdEncoding.EncodeToString(content)
	var b strings.Builder
	for i := 0; i < len(enc); i += 76 {
		j := i + 76
		if j > len(enc) {
			j = len(enc)
		}
		b.WriteString(enc[i:j])
		b.WriteString("\n")
	}
	return []byte(b.String())
}

// FaviconHash FOFA icon_hash：与 Python
// mmh3.hash(codecs.encode(content, "base64")) 逐位一致（含换行语义）。
func FaviconHash(content []byte) int64 {
	return int64(int32(murmur3x8632(fofaBase64(content))))
}

// faviconMaxBytes favicon 采集尺寸上限（超过按未采集处理）。
const faviconMaxBytes = 512 * 1024

// FaviconHashOf 抓取站点默认 favicon（/favicon.ico）并计算 icon_hash。
// 失败/非 200/超限一律返回 0（表示未采集；icon_hash 规则对 0 不匹配）。
func FaviconHashOf(client *httpx.Client, baseURL string) int64 {
	if client == nil || baseURL == "" {
		return 0
	}
	u := strings.TrimRight(baseURL, "/") + "/favicon.ico"
	resp, err := client.GetDirect(u)
	if err != nil || resp == nil || resp.Status != 200 {
		return 0
	}
	if len(resp.Body) == 0 || len(resp.Body) > faviconMaxBytes {
		return 0
	}
	return FaviconHash([]byte(resp.Body))
}
